zjit/src/backend/arm64/mod.rs

```diff
@@ -765,11 +765,36 @@ impl Assembler
                 Target::Label(label_idx) => {
                     // Here we're going to save enough space for ourselves and
                     // then come back and write the instruction once we know the
-                    // offset. We're going to assume we can fit into a single
-                    // b.cond instruction. It will panic otherwise.
-                    cb.label_ref(label_idx, 4, |cb, src_addr, dst_addr| {
-                        let bytes: i32 = (dst_addr - (src_addr - 4)).try_into().unwrap();
-                        bcond(cb, CONDITION, InstructionOffset::from_bytes(bytes));
+                    // offset.
+                    cb.label_ref(label_idx, (cb.conditional_jump_insns() * 4) as usize, |cb, src_addr, dst_addr| {
+                        let reserved_bytes = (cb.conditional_jump_insns() * 4) as i64;
+                        let instruction_addr = src_addr - reserved_bytes;
+                        let bytes: i32 = (dst_addr - instruction_addr).try_into().unwrap();
+                        let offset = InstructionOffset::from_bytes(bytes);
+
+                        let num_insns = if bcond_offset_fits_bits(offset.into()) {
+                            bcond(cb, CONDITION, offset);
+                            1
+                        } else if b_offset_fits_bits(offset.into()) {
+                            // Since we cannot encode the offset in a bcond directly, generate
+                            // an unconditional branch and then jump past it only if the inverse
+                            // of the original condition is met.
+                            bcond(cb, Condition::inverse(CONDITION), 2.into());
+                            b(cb, offset);
+                            2
+                        } else {
+                            // Since we cannot encode the offset into an instruction, load it into
+                            // some number of registers and then call br.
+                            // NOTE: I'm skeptical of the `.try_into().unwrap()`
+                            let load_insns: i32 = emit_load_size(dst_addr.try_into().unwrap()).into();
+                            bcond(cb, Condition::inverse(CONDITION), (load_insns + 2).into());
+                            emit_load_value(cb, Assembler::SCRATCH0, dst_addr.try_into().unwrap());
+                            br(cb, Assembler::SCRATCH0);
+                            load_insns + 2
+                        };
+
+                        assert!(num_insns <= cb.conditional_jump_insns());
+                        for _ in num_insns..cb.conditional_jump_insns() { nop(cb); }
```

Comment:
We don't rewrite label jumps, so we shouldn't need `nop`s here.

Reply:
I'm not sure if I follow. Could you provide an example script that would trigger the problem?